akka-cluster/src/main/scala/akka/cluster/ClusterEvent.scala

```diff
@@ -576,6 +607,33 @@ object ClusterEvent {
     if (newState.overview.reachability eq oldState.overview.reachability) Nil
     else List(ReachabilityChanged(newState.overview.reachability))
 
+  /**
+   * INTERNAL API
+   */
+  @InternalApi
+  private[cluster] def diffTombstones(
+      oldState: MembershipState,
+      newState: MembershipState): immutable.Seq[MemberTombstonesChanged] =
+    if (newState.latestGossip.tombstones == oldState.latestGossip.tombstones) Nil
+    else MemberTombstonesChanged(newState.latestGossip.tombstones.keys.toSet) :: Nil
+
+  /**
+   * INTERNAL API
+   */
+  private[cluster] def publishDiff(oldState: MembershipState, newState: MembershipState, pub: AnyRef => Unit): Unit = {
+    diffMemberEvents(oldState, newState).foreach(pub)
+    diffUnreachable(oldState, newState).foreach(pub)
+    diffReachable(oldState, newState).foreach(pub)
+    diffUnreachableDataCenter(oldState, newState).foreach(pub)
+    diffReachableDataCenter(oldState, newState).foreach(pub)
+    diffLeader(oldState, newState).foreach(pub)
+    diffRolesLeader(oldState, newState).foreach(pub)
+    // publish internal SeenState for testing purposes
+    diffSeen(oldState, newState).foreach(pub)
+    diffReachability(oldState, newState).foreach(pub)
+    diffTombstones(oldState, newState).foreach(pub)
```

Comment:
I think it's important to publish the tombstones before diffMemeberEvents. Otherwise there is a possibility that a Gossip is received with a removed member (in tombstones) and with new incarnation as joining. Then ClusterRemoteWatcher will receive MemberJoined before TombstoneChanged.